scripts/broadcast-pndcp-discovery.nse

```diff
@@ -0,0 +1,449 @@
+      results["MAC address"] = get_mac_addr(macaddr)
+    end
+
+  elseif suboption == PNDCP_SUBOPTION_IP_IP then
+    local dcp_suboption_ip_ip_format = ">I2 I4 I4 I4"
+    if #block >= string.packsize(dcp_suboption_ip_ip_format) then
+      local block_info, ip, netmask, gateway = string.unpack(dcp_suboption_ip_ip_format, block)
+      ip = ipOps.fromdword(ip)
+
+      results["IP Info"] = PNDCP_IP_INFO[block_info]
+      results["IP"] = ip
+      results["Netmask"] = ipOps.fromdword(netmask)
+      results["Gateway"] = ipOps.fromdword(gateway)
+
+      -- Add new target if desired and if IP address looks valid
+      if target.ALLOW_NEW_TARGETS and ip ~= "0.0.0.0" then
+        target.add(ip)
+      end
+    end
+  end
+end
+
+-- Parses a DCP Device block suboption in a DCP identify response
+--
+-- @param suboption type of the suboption
+-- @param block the raw block data to parse
+-- @param results A resulsts table to be filled with the parsed data
+local pndcpParseDeviceBlock = function(suboption, block, results)
+  stdnse.debug1("Parsing device block: Suboption: %u, Data: %s", suboption, stdnse.tohex(block))
+
+  if suboption == PNDCP_SUBOPTION_DEVICE_MANUF then
+    results["Device manufacturer"] = string.sub(block, 3)
+
+  elseif suboption == PNDCP_SUBOPTION_DEVICE_NAMEOFSTATION then
+    results["Name of Station"] = string.sub(block, 3)
+
+  elseif suboption == PNDCP_SUBOPTION_DEVICE_DEV_ID then
+    local dcp_suboption_device_id_format = ">x x I2"
+    if #block >= string.packsize(dcp_suboption_device_id_format) then
+      local vendor_id, device_id = string.unpack(dcp_suboption_device_id_format, block)
+      results["Vendor ID"] = ("0x%04x"):format(vendor_id)
+      results["Device ID"] = ("0x%04x"):format(device_id)
+    end
+
+  elseif suboption == PNDCP_SUBOPTION_DEVICE_DEV_ROLE then
+    local dcp_suboption_device_role_format = ">x x B x"
+    if #block >= string.packsize(dcp_suboption_device_role_format) then
+      local device_role = string.unpack(dcp_suboption_device_role_format, block)
+
+      local device_role_strings = {}
+      if device_role == 0x00 then
+        table.insert(device_role_strings, "None")
+      end
+
+      for flag, name in pairs(PNDCP_DEVICE_ROLES) do
+        if device_role & flag ~= 0 then
+          table.insert(device_role_strings, name)
+        end
+      end
+
+      results["Device Role"] = ("0x%02x (%s)"):format(device_role, table.concat(device_role_strings, ", "))
+    end
+
+  -- NOTE: Contains a list what options/suboptions the device supports,
+  -- no need to parse this explicitly, I think
+  elseif suboption == PNDCP_SUBOPTION_DEVICE_DEV_OPTIONS then
+
+  -- FIXME: I didn't see the following suboptions on any of my devices,
+  -- so I'm not parsing them here as I can't test it
+  elseif suboption == PNDCP_SUBOPTION_DEVICE_ALIAS_NAME then
+  elseif suboption == PNDCP_SUBOPTION_DEVICE_DEV_INSTANCE then
+  elseif suboption == PNDCP_SUBOPTION_DEVICE_OEM_DEV_ID then
```

Comment:
I just added this. And it turns out two of my devices even support suboption `7` and `8`. I am going to investigate this further and see if I can parse them as well.

Reply:
I am parsing all of these options now and have an `else` branch which possibly handles unknown or out-of-spec blocks.
Suboption `7` and `8` I could test, `6` - the "Alias Name" not, but I figured it's just a simple string just like the "Station Name" or "Manufacturer Name", so I added this as well. The wireshark dissector does the same, so that should be alright.